sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1050 @@
+recording state, which allows a `command_graph` object to be populated by the
+command-groups submitted to the queue. Once the graph is complete, recording
+finishes on the queue to put it back into the default executing state. The
+graph is then finalized so that no more nodes can be added. Lastly, the graph is
+submitted as a whole for execution via
+`queue::submit(command_graph<graph_state::executable>)`.
+
+[source, c++]
+----
+  queue q{default_selector{}};
+
+  // New object representing graph of command-groups
+  ext::oneapi::experimental::command_graph graph;
+  {
+    buffer<T> bufferA{dataA.data(), range<1>{elements}};
+    buffer<T> bufferB{dataB.data(), range<1>{elements}};
+    buffer<T> bufferC{dataC.data(), range<1>{elements}};
+
+    // `q` will be put in the recording state where commands are recorded to
+    // `graph` rather than submitted for execution immediately.
+    q.begin_recording(graph);
+
+    // Record commands to `graph` with the following topology.
+    //
+    //      increment_kernel
+    //       /         \
+    //   A->/        A->\
+    //     /             \
+    //   add_kernel  subtract_kernel
+    //     \             /
+    //   B->\        C->/
+    //       \         /
+    //     decrement_kernel
+
+    q.submit([&](handler &cgh) {
+      auto pData = bufferA.get_access<access::mode::read_write>(cgh);
+      cgh.parallel_for<increment_kernel>(range<1>(elements),
+                                         [=](item<1> id) { pData[id]++; });
+    });
+
+    q.submit([&](handler &cgh) {
+      auto pData1 = bufferA.get_access<access::mode::read>(cgh);
+      auto pData2 = bufferB.get_access<access::mode::read_write>(cgh);
+      cgh.parallel_for<add_kernel>(range<1>(elements),
+                                   [=](item<1> id) { pData2[id] += pData1[id]; });
+    });
+
+    q.submit([&](handler &cgh) {
+      auto pData1 = bufferA.get_access<access::mode::read>(cgh);
+      auto pData2 = bufferC.get_access<access::mode::read_write>(cgh);
+      cgh.parallel_for<subtract_kernel>(
+          range<1>(elements), [=](item<1> id) { pData2[id] -= pData1[id]; });
+    });
+
+    q.submit([&](handler &cgh) {
+      auto pData1 = bufferB.get_access<access::mode::read_write>(cgh);
+      auto pData2 = bufferC.get_access<access::mode::read_write>(cgh);
+      cgh.parallel_for<decrement_kernel>(range<1>(elements), [=](item<1> id) {
+        pData1[id]--;
+        pData2[id]--;
+      });
+    });
+
+    // queue will be returned to the executing state where commands are
+    // submitted immediately for extension.
+    q.end_recording();
+  }
+
+  // Finalize the modifiable graph to create an executable graph that can be
+  // submitted for execution.
+  auto exec_graph = graph.finalize(q.get_context());
+
+  // Execute graph
+  q.submit(exec_graph);
+----
+
+== Issues
+
+=== Multi Device Graph
+
+Allow an executable graph to contain nodes targeting different devices.
+
+**Outcome:** Under consideration
```

Comment:
We are seeing highly heterogeneous customer devices where each device/chip contains multiple, different programmable accelerator units - which can be considered independent devices. To target such next-gen chips we need support for multi-device graphs. I will also open a related issue.

Reply:
Where is the best place to file such an issue or open such a discussion, e.g., is there a way to make such an issue SYCL-Graph specific?

Reply:
For now, we organize SYCL-graph related issues here:
https://github.com/reble/llvm/issues
Regarding multi-device graph support there is already an open issue:
https://github.com/reble/llvm/issues/7

Reply:
reble#7 would indeed be the place to put your thoughts.

Would caveat that by saying that the issue tracker has so far been used to coordinate tasks rather than solicit external requests. So in general, I would prefer discussions are started here to make the thread visible to all reviewers and then taken to issues if necessary, rather than going straight to the issue tracker.

Reply:
Thank you, added to the issue as a comment here: https://github.com/reble/llvm/issues/7#issuecomment-1380915706